garage/tf/policies/epsilon_greedy_policy.py

```diff
@@ -0,0 +1,118 @@
+"""
+ϵ-greedy policy.
+
+Random exploration according to the value of epsilon.
+"""
+import numpy as np
+
+from garage.core import Serializable
+from garage.misc.overrides import overrides
+from garage.tf.policies import Policy
+from garage.tf.spaces import Box
+
+
+class EpsilonGreedyPolicy(Policy, Serializable):
+    """
+    ϵ-greedy policy.
+
+    Select action based on the value of ϵ. ϵ will decrease from
+    max_epsilon to min_epsilon within decay_ratio * total_step.
+
+    At state s, with probability
+    1 − ϵ: select action = argmax Q(s, a)
+    ϵ    : select a random action
+
+    Args:
+        env_spec: environment specification
+        total_step: Total steps in the training.
+        max_epsilon: the maximum(starting) value of epsilon.
+        min_epsilon: the minimum(terminal) value of epsilon.
+        decay_ratio: Fraction of total steps for epsilon decay.
+    """
+
+    def __init__(self,
+                 env_spec,
+                 total_step,
+                 max_epsilon=1.0,
+                 min_epsilon=0.02,
+                 decay_ratio=0.1):
+        Serializable.quick_init(self, locals())
+        super(EpsilonGreedyPolicy, self).__init__(env_spec)
+
+        self._env_spec = env_spec
+        self._max_epsilon = max_epsilon
+        self._epsilon = max_epsilon
+        self._min_epsilon = min_epsilon
+        self._decay_period = int(total_step * decay_ratio)
+        self._action_space = env_spec.action_space
+        self._act = None
+
+    def set_act(self, act):
```

Comment:
This is not how things are usually designed in garage.

The Q-Function should be set during the constructor.

I believe what you actually want here is an ExplorationStrategy which wraps a Q-Function-derived policy of your choice.

I am not opposed to changing the inheritance tree to make ExplorationStrategy inherit from/wrap policies (which is basically what you did here). If you want to make that change, you should make it consistent across the codebase.

Reply:
Yes, I want a ExplorationStrategy which wraps a Q-Function-derived policy. The reason I pass the Q-Function here is that I won't have a Q-Function built until I constructed the algorithm class, i.e. `DQN` in this case. However, ExplorationStrategy is one of the argument needed. I can only either set the Q-Function in the ExplorationStrategy or set the ExplorationStrategy in the DQN class afterwards.

Reply:
I will try to use ExplorationStrategy class and pass q-values to `get_actions`.